[Jed] Step 5-2 : 리팩토링

junu0516 opened the pull request:
### 작업 내용
- [x] Player 클래스 내에서 PokerGame의 하위타입을 참조하던 부분 수정
- [x] 카드를 빼는 행위가 정의된 takeOutCard 메소드를 Player에서 자식클래스인 Dealer로 이동
- [x] 사실상 불필요한 행위인  Dealer 클래스 내부의 takeOutAllNecessaryCards 메소드 제거
- [x] CardDeck 구조체를 클래스로 변경 후, PokerGame과 Dealer 클래스가 참조값을 공유하도록 수정
    
### 고민과 해결
#### 1. Player 클래스 내에서 PokerGame.Count 참조타입을 사용하던 부분 수정
- PokerGame.Count가 아닌 Int 타입을 사용하도록 수정했습니다.
- PokerGame에서 Player을 초기화할 때도 Player.Count.rawValue를 넘기도록 수정했습니다.
- 어차피 PokerGame이라는 외부에서 열거형을 통해 정해진 Int값을 사용하기 때문에, Player 생성자 파라미터에서 Int 타입으로 받더라도 미리 정해진 case 중 하나만 들어올 것이므로 플레이어의 수를 정의하기 위한 별도의 타입 선언을 하지 않아도 문제 없다고 판단했습니다! 
    
#### 2. PokerGame 시작 시, 딜러가 카드 덱에서 카드를 가져와서 플레이어에게 나눠주던 로직을 좀 더 간소화
- 기존에는 카드 덱에 있던 52장의 카드 중, 플레이어에게 나눠줄 모든 카드를 우선 딜러가 가져온 후 다시 여기서 플레이어에게 1장씩 나눠주는 식으로 구현했습니다.
- 하지만 어차피 결과적으로 덱에 있는 카드를 1장씩 나눠주는 행위인데 굳이 딜러가 나눠줄 카드를 우선 가져오는 행위가 비효율적이라 판단했습니다.
- 따라서 위의 행위를 수행하던 takeOutAllNecessAryCards라는 Dealer 내부에 있던 메소드를 우선 제거한 후, Player 클래스에 있던 카드 덱에서 카드를 한장씩 빼오는 takeOutCard 메소드를 Dealer로 옮겨서, PokerGame의 distributeCards() 메소드 내부에서는 결과적으로 Dealer.takeOutCard()가 리턴하는 카드를 플레이어에게 넘기는 식으로 수정했습니다.
    - 이를 위해 또한 PokerGame과 Dealer이 공통으로 CardDeck을 관리하는 것이 낫겠나고 판단했습니다.
    - CardDeck을 구조체로 구현했던 것을 클래스로 변경한 후 Dealer에 CardDeck 타입 프로퍼티를 추가해서 PokerGame과 Dealer이 공통의 참조값을 바라보도록 수정했습니다.
    
#### 3.Player.cards 프로퍼티 완전한 private로 변경
- 다시 살펴보니 cards 프로퍼티가 private (set)으로 되어 있어서 private으로 수정했습니다.
- 외부에서 Player 내부의 카드의 수를 가져와야 할 필요가 있기에 별도로 카드 배열의 길이를 리턴하는 cardsCount 프로퍼티를 선언해서 이걸 사용하도록 수정했습니다.

junu0516 committed:
Player.cards 프로퍼티 private으로 변경

- private (set) -> private

junu0516 committed:
CardDeck 구조체에서 클래스로 변경

- PokerGame, Dealer 이 공통으로 CardDeck 참조를 공유하기 위해 구조체에서 클래스로 변경

junu0516 committed:
Dealer 클래스 내부에 CardDeck 타입 프로퍼티 추가

- PokerGame이 가지고 있는 CardDeck 타입 인스턴스의 참조값을 공유

junu0516 committed:
Player 클래스 변경

- Players.takeOutCard() 메소드 제거(자식 클래스로 이동)
- 플레이어가 가진 카드 순회를 위한 getCardsIterator() 메소드 추가

junu0516 committed:
Dealer 클래스 수정

- 기존 Player 부모 클래스에 있던 takeOutCard 메소드 Dealer 자식 클래스로 이동
- 사용하지 않을 takeOutAllNecessaryCards() 메소드 제거

junu0516 committed:
Players 클래스 수정

- 사용하지 않을 addCard() 메소드 제거

junu0516 committed:
PokerGame 클래스 수정

- 기존에 딜러가 나눠줄 카드를 모두 가져가던 takeOutAllNecessaryCards() 메소드 제거
- distributeCards() 메소드 내부에서 dealer이 추가된 players의 이터레이터를 받아서 카드 분배하도록 수정
- dealer 내부의 CardDeck에서 카드를 takeOut()메소드 호출을 통해 한장씩 빼서 dealer이 포함된 이터레이터 내의 개별 player에 카드를 한장씩 추가

junu0516 committed:
ViewController 클래스 수정

- player.cards가 private이기 때문에 직접 접근해서 이터레이터를 가져오지 않고, player 내부적으로 이터레이터를 받아서 리턴하는 메소드를 호출

junu0516 committed:
Player 클래스 수정

- Players.card가 private이기 때문에, 플레이어가 가진 카드의 수를 리턴하는 cardsCount 프로퍼티를 별도로 추가

junu0516 committed:
PokerGameTest 테스트 클래스 수정

- 플레이어의 카드 수를 가져올 때 player.cards에 직접 접근 불가능하기 때문에 추가한 cardsCount 프로퍼티값을 읽어와서 테스트

junu0516 committed:
Players 클래스 수정

- Players.numberOfPlayers 프로퍼티 제거
- Players 생성자의 파라미터로 PokerType.Count 타입을 받지 말고 외부에서 값이 결정된 Int타입으로 받도록 수정
- Player.createPlayers() 메소드의 파라미터로 생성자 파라미터로 받은 Int 타입 numberOfPlayers 값을 받도록 수정

junu0516 committed:
PokerGame 클래스 수정

- 생성자에서 self.players 초기화할 때 파라미터 타입으로 Int타입 넘기도록 수정

junu0516 commented:
> 1. Player 클래스 내에서 PokerGame.Count 참조타입을 사용하던 부분 수정


이 부분 관련해서 PR 보내놓고 다시 생각해보니, Int값을 그대로 사용하는 것이 여전히 꺼림직해서 다른 방식으로 수정하고자 합니다.
 플레이어 수에 해당하는 값은 createPlayers() 메소드에서 사용하는데 차라리 PokerGame 클래스에서 PokerGame.Count 값만큼 Player을 만드는 로직이 담긴 클로저를 만들어서 이걸 Players로 넘겨서 거기서 클로저를 실행하는 방식이 더 낫지 않을까 싶어 수정중입니다!

 - Player 생성의 역할은 결과적으로 Players 내부에서 가지게 됨
 - PokerGame.Count 관련 값을 Players 내부에서 가져가 사용할 필요가 없어짐

위와 같은 이유로 클로저를 넘기는 방식을 생각했습니다.

junu0516 committed:
Players 생성자 수정

- Player 배열을 생성하는 클로저를 파라미터로 받아서 내부에서 클로저 실행해서 players 배열 초기화

junu0516 committed:
PokerGame 생성자 수정

- 생성자에서 Player 배열 만드는 클로저 작성해서 Players 생성자 파라미터로 넘기도록 수정

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -19,8 +22,8 @@ class Player: CustomStringConvertible{
-        guard let card = self.cards.removeOne() else { return nil }
-        return card
+    func getCardsIterator() -> IndexingIterator<[Card]>{
+        return self.cards.getIterator()
```
앞으로는 하위 타입이 값만 있는 경우에 그 값이 많이 들어갈 수 있는 경우에는
이터레이터를 제공해주는 시도는 좋습니다

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -1,39 +1,24 @@
+    init(creatingLoop: ()->[Player]){
+        for player in creatingLoop(){
+            self.players.append(player)
         }
```
생성자에서 클로저가 어떤 역할을 하는 건가요? 꽤 어려운 생성자네요. 
player 객체 생성의 역할을 외부에서 하고 여기서는 배열에 추가만 하려는 거라면 `init(players: [Player])` 형태와 크게 다른 점이 없네요. 오히려 복잡해지기만 했죠
Player 객체를 누가 생성할꺼냐 기준을 정해보면 좋겠습니다. 그런 경우 Player에 필요한 속성을 가장 잘 알고 있는 타입이면 좋습니다

junu0516 replied:
원래는 PokerGame.Count.rawValue를 생성자의 파라미터로 넘겼는데, 이렇게 될 경우에는 위에 언급한대로 Player의 생성자 파라미터 타입이 Int가 되는 점이 좀 걸렸습니다.

그래서 Int값을 넘기는 코드를 작성하지 않으려고 외부에서 PokerGame.Count만 사용하도록 하려 한건데 결과적으로 Players의 존재의미만 약하게 만들어버린 것 같습니다.

Player 배열 초기화 역할은 기존처럼 Players 내부에서 수행하도록 하고 생성자의 파라미터로 PokerGame.Count.rawValue를 넘기는 PokerGame.Count를 활용할 수 있는 다른 방안을 생각해보겠습니다.

godrm replied:
PokerGame.Count 타입으로 클로저를 넘기고 루프를 제공하는 방식도 고려해보세요. 그러면 값을 가져오지 않아도 될 겁니다.

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -1,39 +1,24 @@
+            
     func addPlayer(_ player: Player){
         self.players.append(player)
     }
```
이렇게 addPlayer가 있으면 init()에서 player를 추가할 필요가 없는 것처럼 보이기도 하네요

junu0516 replied:
아 이부분은 기존에 Players 내부에서  Player 배열을 초기화하면서 쓰던건데 수정하면서 깜빡하고 제거하지 않은 것 같습니다. 기존처럼 배열을 초기화하는 역할을 Players 내부에서 수행하도록 할 것이기 때문에 그대로 두겠습니다.

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -24,9 +23,18 @@ class PokerGame: CustomStringConvertible{
+
+        var randomNames = ["dale","eddy","jee","foucault","sol"]
+        self.players = Players(creatingLoop: {
+            var players: [Player] = []
+            for _ in 0..<numberOfPlayers.rawValue{
+                guard let playerName = randomNames.popLast() else { continue }
+                players.append(Player(name: playerName))
+            }
+            return players
```
이 클로저는 Players 타입에서 해야할 일을 상위가 다 정해주는 방식이네요. 역할 분리가 된게 아니라 Players가 빈약한 형태인 것 같습니다.

junu0516 replied:
음 말씀하신대로 Player 객체를 생성하는 역할은 Players 내부에서 온전히 수행하는 것이 맞는 것 같습니다

godrm commented:
`Player 생성의 역할은 결과적으로 Players 내부에서 가지게 됨` 이 부분은 코드를 보면 
Players가 하는 게 없는 것 같습니다. 상위에서 클로저로 다 생성해주고 있는 게 아닌가요?

godrm commented:
머지하고 마무리하실 건가요? 더 개선해보실래요? 확인해서 댓글 남겨주세요

junu0516 commented:
> `Player 생성의 역할은 결과적으로 Players 내부에서 가지게 됨` 이 부분은 코드를 보면 Players가 하는 게 없는 것 같습니다. 상위에서 클로저로 다 생성해주고 있는 게 아닌가요?

클로저를 생성자의 파라미터로 넘겨서, 생성자 내부에서 클로저를 실행하기 때문에 생성의 역할을 Players가 수행한다고 생각했는데 이미 클로저를 생성하는 시점부터 사실상 PokerGame에서 해당 역할을 수행하는 것이기 때문에 제가 잘못 생각한 것 같습니다!

위에 언급한대로 원래는 PokerGame.Count.rawValue를 넘겨서 Players의 생성자가 __`init(numberOfPlayers: Int)`__ 의 형태가 되는 것을 막고자 한 건데, 우선 클로저를 넘기는 것 대신 PokerGame.Count 형태를 활용할 수 있는 다른 방법을 마지막으로 생각해보려 합니다.

junu0516 commented:
> 머지하고 마무리하실 건가요? 더 개선해보실래요? 확인해서 댓글 남겨주세요

피드백주신 부분 반영해서 좀 더 개선해보겠습니다.

godrm commented:
그럼 PR 열어놓을테니 개선해서 Push하고 댓글 남겨주세요

junu0516 committed:
Players 클래스 수정

- 기존의 PokerGame.Count 내부 타입을 Players.Count로 이동&변경
- Players.Count 열거형 내에 원하는 클로저를 반복 실행하는 creatingLoop() 메소드 추가
- Players 생성자는 Players.Count 타입을 파라미터로 받아서 결과적으로 creatingLoop() 호출하여 배열 초기화

junu0516 committed:
PokerGame 클래스 수정

- 기존에 가지고 있던 PokerGame.Count 내부타입 및 관련 로직 모두 제거
- 생성자 파라미터로 Players.Count 타입을 받아서, Players 인스턴스 생성 시 Players의 생성자 파라미터로 그대로 넘김

junu0516 committed:
ViewController 클래스 수정

- 기존 수정내역에 맞춰서 Players.Count로 모두 수정

junu0516 committed:
PokerGameTest 테스트 클래스 수정

- 기존 수정내역에 맞춰서 Players.Count로 모두 수정

junu0516 committed:
Player 클래스 수정

- 다른 클래스들과 맞춰서 맨 위 공백 제거

junu0516 commented:
> PokerGame.Count 타입으로 클로저를 넘기고 루프를 제공하는 방식도 고려해보세요. 그러면 값을 가져오지 않아도 될 겁니다.

제안해주신 부분 반영해서 수정했습니다!
- 기존의 PokerGame.Count 로 선언한 내부 타입을 Players.Count로 Players 클래스의 하위 타입으로 이동&변경했습니다.
- 이후 외부에서 PokerGame 인스턴스를 생성할 때는 Players.Count 타입을 생성자 파라미터로 전달하고, PokerGame 생성자 내부에서 Players 인스턴스를 생성할 때 Players의 생성자 파라미터로 Players.Count 타입을 그대로 넘기도록 했습니다.
- Players 내부에서는 우선 Players.Count 열거형 내에 파라미터로 전달받은 클로저를 반복실행하는 __`creatingLoop()`__ 함수를 정의해서, 결과적으로 Players 생성자에서 해당 함수를 호출하여 Player 배열을 초기화할 수 있도록 했습니다.
    - (말씀하신 클로저 넘겨서 루프 실행하는게 혹시 이 방식을 말씀하신걸까요?) 
```swift
class Players{
    internal enum Count: Int, CaseIterable{
        case two = 2
        case three = 3
        case four = 4
        
        func creatingLoop(loop: ()->Void){
            for _ in 0..<self.rawValue{
                loop()
            }
        }
    }
   
    private var players: [Player]
   
    init(numberOfPlayers: Count){
        createPlayers(numberOfPlayers: numberOfPlayers)
    }
    
    func createPlayers(numberOfPlayers: Count){
        numberOfPlayers.creatingLoop {
            guard let playerName = randomNames.popLast() else { return }
            players.append(Player(name: playerName))
        }
    }
}
```

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -1,6 +1,20 @@
+                loop()
+            }
+        }
+    }
```
이게 지금은 너무 작은 역할만 하는 타입이지만, 이런 식으로 내부 값을 가져가지 않고 활용할 수 있다는 점을 기억하시면 됩니다.

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -9,9 +23,14 @@ class Players: CustomStringConvertible{
+        createPlayers(numberOfPlayers: numberOfPlayers)
+    }
+    
+    func createPlayers(numberOfPlayers: Count){
```
내부에서만 쓰는 메소드라면 private func 으로 선언해주세요

godrm commented on PokerGameApp/PokerGameApp/PokerGame.swift:
```diff
@@ -21,20 +15,11 @@ class PokerGame: CustomStringConvertible{
     }
     
-    init(numberOfPlayers: Count, stud: Stud){
+    init(numberOfPlayers: Players.Count, stud: Stud){
```
이 지점에서 Players 타입을 PokerGame 외부에서 알 필요가 없다면, Players.Count 를 PokerGame.Count로 만들어도 됩니다. 
그렇게 어느 정도 수준에서 필요한가로 하위 타입을 선언해도 됩니다. 숨길건 숨기고 공개할 건 공개하고 하는거죠

godrm reviewed (approved):
수고하셨습니다.
